tenbilac/layer.py

```diff
@@ -16,49 +16,104 @@
 
 		
 class Layer():	
-	def __init__(self, ni, nn, actfct=act.tanh, name="None"):
+	def __init__(self, ni, nn, actfct=act.tanh, name="None", mode="sum"):
 		
 		"""
 		:param ni: Number of inputs
 		:param nn: Number of neurons
 		:param actfct: Activation function
+		:param mode: "sum" or "mult". If sum, the neurons perform the usual linear combinations.
+			If mult, they work as "Product Units" (Durbin Rumelhart 1989).
 		"""
 
-		self.ni = ni
-		self.nn = nn
+		self.ni = abs(ni)
+		self.nn = abs(nn) # We keep those positive, even if negative might have indicated "mult" earlier in the code.
 		self.actfct = actfct
 		self.name = name
 		
+		self.mode = mode
+		if not self.mode in ("sum", "mult"):
+			raise RuntimeError("Unknown layer mode")
+
 		self.weights = np.zeros((self.nn, self.ni)) # first index is neuron, second is input
 		self.biases = np.zeros(self.nn) # each neuron has its bias
+			
 		
+	def __str__(self):
+		return "Layer '{self.name}', mode {self.mode}, ni {self.ni}, nn {self.nn}, actfct {self.actfct.__name__}".format(self=self)
 	
-	def addnoise(self, wscale=0.1, bscale=0.1):
+	def modecode(self):
+		"""
+		Just a little "*" to label unconventional layers in logs etc
+		"""
+		if self.mode == "sum":
+			return ""
+		else:
+			return "*"
+	
+	def addnoise(self, wscale=0.1, bscale=0.1, multwscale=0.1):
 		"""
 		Adds some noise to weights and biases
 		"""
-		self.weights += wscale * np.random.randn(self.weights.size).reshape(self.weights.shape)
+		if self.mode == "sum":
+			self.weights += wscale * np.random.randn(self.weights.size).reshape(self.weights.shape)
+		elif self.mode == "mult":
+			self.weights += multwscale * np.random.randn(self.weights.size).reshape(self.weights.shape)
+		else:
+			raise RuntimeError("Unknown mode")
 		self.biases += bscale * np.random.randn(self.biases.size)
 	
-	def zero(self):
+	def setzero(self):
 		"""
 		Sets all weights and biases to zero
 		"""
-		logger.info("Setting layer '{self.name}' parameters to zero...".format(self=self))
+		logger.info("Setting {self.mode}-layer '{self.name}' parameters to zero...".format(self=self))
+		#self.weights = np.zeros(self.weights.shape) # No, this does not work! It kills the get_params reference used to set up the Training *before* noise is added.
+		#self.biases = np.zeros(self.biases.shape) # Leave this commented to prevent later attempts to change it.
 		self.weights *= 0.0
 		self.biases *= 0.0
 	
+
+	def setidentity(self, onlyn=None):
+		"""
+		Sets the weights and biases so that this layer transports all its first nn inputs to the nn outputs.
+		This is meant to give a promising initial condition for training a calibration-like task. 
+		
+		:param onlyn: If set, limits the number of neurons to be set. By default, all are set.
```

Comment:
Why are you setting only the first n instead of passing which weights should be set to 1?